File: code/packages/csharp/ct-compare/CtCompare.cs
namespace CodingAdventures.CtCompare;

/// <summary>
/// Constant-time comparison helpers for public-length byte buffers.
/// </summary>
public static class CtCompare
{
    /// <summary>
    /// Compare two byte sequences without short-circuiting on the first differing byte.
    /// </summary>
    public static bool CtEq(ReadOnlySpan<byte> left, ReadOnlySpan<byte> right)
    {
        if (left.Length != right.Length)
        {
            return false;
        }

        byte accumulator = 0;
        for (var i = 0; i < left.Length; i++)
        {
            accumulator |= (byte)(left[i] ^ right[i]);
        }

        return accumulator == 0;
    }

    /// <summary>
    /// Compare two same-size byte arrays. Runtime length mismatch returns false.
    /// </summary>
    public static bool CtEqFixed(ReadOnlySpan<byte> left, ReadOnlySpan<byte> right) => CtEq(left, right);

    /// <summary>
    /// Select between two same-length byte arrays using a byte mask.
    /// </summary>
    public static byte[] CtSelectBytes(ReadOnlySpan<byte> left, ReadOnlySpan<byte> right, bool choice)
    {
        if (left.Length != right.Length)
        {
            throw new ArgumentException("CtSelectBytes requires equal-length inputs.");
        }

        var mask = unchecked((byte)(0 - (choice ? 1 : 0)));
        var output = new byte[left.Length];
        for (var i = 0; i < left.Length; i++)
        {
            output[i] = (byte)(right[i] ^ ((left[i] ^ right[i]) & mask));
        }

        return output;
    }

    /// <summary>
    /// Compare two UInt64 values without data-dependent ordering branches.
    /// </summary>
    public static bool CtEqUInt64(ulong left, ulong right)
    {
        var diff = left ^ right;
        var folded = (diff | (0UL - diff)) >> 63;
        return folded == 0;
    }
}

File: code/packages/fsharp/ct-compare/CtCompare.fs
namespace CodingAdventures.CtCompare

/// Constant-time comparison helpers for public-length byte buffers.
[<RequireQualifiedAccess>]
module CtCompare =
    let ctEq (left: byte array) (right: byte array) =
        if isNull left then
            nullArg (nameof left)
        if isNull right then
            nullArg (nameof right)

        if left.Length <> right.Length then
            false
        else
            let mutable accumulator = 0uy
            for index in 0 .. left.Length - 1 do
                accumulator <- accumulator ||| (left[index] ^^^ right[index])
            accumulator = 0uy

    let ctEqFixed (left: byte array) (right: byte array) =
        ctEq left right

    let ctSelectBytes (left: byte array) (right: byte array) choice =
        if isNull left then
            nullArg (nameof left)
        if isNull right then
            nullArg (nameof right)

        if left.Length <> right.Length then
            invalidArg (nameof right) "ctSelectBytes requires equal-length inputs."

        let mask = if choice then 0xFFuy else 0uy
        Array.init left.Length (fun index ->
            right[index] ^^^ ((left[index] ^^^ right[index]) &&& mask))

    let ctEqUInt64 (left: uint64) (right: uint64) =
        let diff = left ^^^ right
        let folded = (diff ||| (0UL - diff)) >>> 63
        folded = 0UL
